knit/core.py

```diff
@@ -248,6 +263,7 @@ def preexec_func():
             # preexec_fn not supported on Windows
             proc = Popen(args, stdin=PIPE)
         self.proc = proc
+        atexit.register(self.__del__)
```

Comment:
I don't think this is what you want to do, as `__del__` will be called twice for every object then. Instead the following might be better:

- On `__init__` add the `knit` object to a `weakref.WeakSet`
- Add a `atexit.register` call to a function that will iterate over this set at close and remove any lingering instances.

The only reason there would be lingering objects is if there is a reference cycle, which leads to no `__del__` in the cycle being called. These objects will still be in the weakset, allowing the exit handler to finish them up.

Reply:
It should be enough to set `self.app_id = None`, no? del does nothing after that happens.
The error messages you sometimes get come from py4j's watcher thread, which communicates with the java process.

Reply:
Registering the method will hold on to a reference to the object, preventing actual cleanup until exit in all cases.